src/thor/triplegbuilder.cc

```diff
@@ -1095,25 +1056,24 @@ TripLegBuilder::Build(const AttributesController& controller,
         ++begin_index;
       }
 
-    } else if (is_first_edge || is_last_edge) {
-      // We need to clip the shape if i its at the beginning or end and
-      // is not full length
-      float length = std::max(static_cast<float>(directededge->length()) * length_pct, 1.0f);
-      if (directededge->forward() == is_last_edge) {
-        AddPartialShape<std::vector<PointLL>::const_iterator>(trip_shape, edgeinfo.shape().begin(),
-                                                              edgeinfo.shape().end(), length,
-                                                              is_last_edge,
-                                                              is_last_edge ? end_vrt : start_vrt);
-      } else {
-        AddPartialShape<std::vector<PointLL>::const_reverse_iterator>(trip_shape,
-                                                                      edgeinfo.shape().rbegin(),
-                                                                      edgeinfo.shape().rend(), length,
-                                                                      is_last_edge,
-                                                                      is_last_edge ? end_vrt
-                                                                                   : start_vrt);
+    } // We need to clip the shape if its at the beginning or end
+    else if (is_first_edge || is_last_edge) {
+      // Get edge shape and reverse it if directed edge is not forward.
+      auto edge_shape = edgeinfo.shape();
+      if (!directededge->forward()) {
+        std::reverse(edge_shape.begin(), edge_shape.end());
       }
-    } else {
-      // Just get the shape in there in the right direction
+      // We need to clip the shape if its at the beginning or end and is not full length
+      float total = static_cast<float>(directededge->length());
+      // Trim the shape
+      TrimShape(edge_shape, is_first_edge ? start_pct * total : 0,
+                is_first_edge ? start_vrt : edge_shape.front(),
+                is_last_edge ? end_pct * total : total, is_last_edge ? end_vrt : edge_shape.back());
```

Comment:
For readability can you pull the ternaries out of the function call and name them?